### PR TITLE
LLVM and SPIRV-LLVM-Translator pulldown (WW18)

### PR DESCRIPTION
**Pull request by vmaksimo**
LLVM: llvm/llvm-project@fcb45b5
SPIRV-LLVM-Translator: KhronosGroup/SPIRV-LLVM-Translator@a9d3819

### CODE REVIEW

**Comment by vmaksimo** [comment]
Build on CUDA failed with the following error:
```
Stack dump:
0. Program arguments: /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./bin/opt -O3 -o /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./lib/clc/builtins.opt.libspirv-nvptx64--.bc /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./lib/clc/builtins.link.libspirv-nvptx64--.bc
1. Running pass 'Function Pass Manager' on module '/localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./lib/clc/builtins.link.libspirv-nvptx64--.bc'.
2. Running pass 'SLP Vectorizer' on function '@_Z17__spirv_ocl_crossDv3_DF16_S_'
```
Looks like the problem could be in SLP Vectorither or in libclc (this intrinsic is defined in [libclc/generic/libspirv/geometric/cross.cl](https://github.com/intel/llvm/blob/sycl/libclc/generic/libspirv/geometric/cross.cl)).
@bader could you please suggest the right person to investigate this failure?

**Comment by bader** [comment]
> Build on CUDA failed with the following error:
> 
> ```
> Stack dump:
> 0. Program arguments: /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./bin/opt -O3 -o /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./lib/clc/builtins.opt.libspirv-nvptx64--.bc /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./lib/clc/builtins.link.libspirv-nvptx64--.bc
> 1. Running pass 'Function Pass Manager' on module '/localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/./lib/clc/builtins.link.libspirv-nvptx64--.bc'.
> 2. Running pass 'SLP Vectorizer' on function '@_Z17__spirv_ocl_crossDv3_DF16_S_'
> ```
> Looks like the problem could be in SLP Vectorither or in libclc (this intrinsic is defined in [libclc/generic/libspirv/geometric/cross.cl](https://github.com/intel/llvm/blob/sycl/libclc/generic/libspirv/geometric/cross.cl)).

@vmaksimo, do you know which particular change caused this issue?

@Naghasan, @steffenlarsen, @turinevgeny, FYI.

**Comment by vmaksimo** [comment]
Actually, there was a bunch of changes to SLP that are included in this PR, so it's hard to say which was the guilty one:
```
| | * 18c61fc498c7 2021-04-22 | [SLP]Skip undefs trying to find perfect/shuffled tree entries matching. [Alexey Bataev]
| | * d4f5f23bbbe5 2021-04-22 | [SLP]Replace more `TTI` with `TTIRef`, NFC. [Alexey Bataev]
| | * da2cdfd4211a 2021-04-22 | [SLP]Added explicit ref to TargetTransformInfo to try to pacify MSVC buildbots, NFC. [Alexey Bataev]
| | * e99b98cb1bca 2021-04-06 | [SLP]Improve cost model for the vectorized extractelements. [Alexey Bataev]
| | * 07c236f3c3fa 2021-04-21 | [SLP]Add a test with broadcast shuffle kind in SLP, NFC. [Alexey Bataev]
| | * | | af870e11aed7 2021-04-20 | [SLP] Add detection of shuffled/perfect matching of tree entries. [Alexey Bataev]
| | * | | b82344a01949 2021-04-20 | Revert "[SLP] Add detection of shuffled/perfect matching of tree entries." [Alexey Bataev]
| | * | | daf6e18c55c2 2021-04-20 | [SLP] Add detection of shuffled/perfect matching of tree entries. [Alexey Bataev]
| | * | | cf00cb8bed72 2021-04-20 | Revert "[SLP] Add detection of shuffled/perfect matching of tree entries." [Alexey Bataev]
| | * | | b232771acad6 2021-04-20 | [SLP] Add detection of shuffled/perfect matching of tree entries. [Alexey Bataev]
| | * | | 803048106533 2021-04-19 | Revert "[SLP]Add detection of shuffled/perfect matching of tree entries." [Alexey Bataev]
| | * | | d6fde913790d 2021-04-14 | [SLP]Add detection of shuffled/perfect matching of tree entries. [Alexey Bataev]
| | * | | | | | | 72142b909d63 2021-04-14 | [SLP]Added a tests for shuffled matched tree entries, NFC. [Alexey Bataev]
| | * | | | | | | b49c41afbaa2 2021-04-14 | [SLP] createOp - fix null dereference warning. NFCI. [Simon Pilgrim]
```

**Comment by bader** [comment]
> Actually, there was a bunch of changes to SLP that are included in this PR, so it's hard to say which was the guilty one:
> 
> ```
> | | * 18c61fc498c7 2021-04-22 | [SLP]Skip undefs trying to find perfect/shuffled tree entries matching. [Alexey Bataev]
> | | * d4f5f23bbbe5 2021-04-22 | [SLP]Replace more `TTI` with `TTIRef`, NFC. [Alexey Bataev]
> | | * da2cdfd4211a 2021-04-22 | [SLP]Added explicit ref to TargetTransformInfo to try to pacify MSVC buildbots, NFC. [Alexey Bataev]
> | | * e99b98cb1bca 2021-04-06 | [SLP]Improve cost model for the vectorized extractelements. [Alexey Bataev]
> | | * 07c236f3c3fa 2021-04-21 | [SLP]Add a test with broadcast shuffle kind in SLP, NFC. [Alexey Bataev]
> | | * | | af870e11aed7 2021-04-20 | [SLP] Add detection of shuffled/perfect matching of tree entries. [Alexey Bataev]
> | | * | | b82344a01949 2021-04-20 | Revert "[SLP] Add detection of shuffled/perfect matching of tree entries." [Alexey Bataev]
> | | * | | daf6e18c55c2 2021-04-20 | [SLP] Add detection of shuffled/perfect matching of tree entries. [Alexey Bataev]
> | | * | | cf00cb8bed72 2021-04-20 | Revert "[SLP] Add detection of shuffled/perfect matching of tree entries." [Alexey Bataev]
> | | * | | b232771acad6 2021-04-20 | [SLP] Add detection of shuffled/perfect matching of tree entries. [Alexey Bataev]
> | | * | | 803048106533 2021-04-19 | Revert "[SLP]Add detection of shuffled/perfect matching of tree entries." [Alexey Bataev]
> | | * | | d6fde913790d 2021-04-14 | [SLP]Add detection of shuffled/perfect matching of tree entries. [Alexey Bataev]
> | | * | | | | | | 72142b909d63 2021-04-14 | [SLP]Added a tests for shuffled matched tree entries, NFC. [Alexey Bataev]
> | | * | | | | | | b49c41afbaa2 2021-04-14 | [SLP] createOp - fix null dereference warning. NFCI. [Simon Pilgrim]
> ```

Based on this list, @alexey-bataev might be the right person to investigate this crash.

File: llvm/test/Transforms/InstCombine/trunc-extractelement-spir.ll [hunk 1]
```diff
@@ -5,8 +5,8 @@ target triple = "spir"
 
 define i32 @shrinkExtractElt_i64_to_i32_0(<3 x i64> %x) {
 ; CHECK-LABEL: @shrinkExtractElt_i64_to_i32_0(
-; CHECK-NOT:    [[TMP1:%.*]] = bitcast <3 x i64> [[X:%.*]] to <6 x i32>
-; CHECK-NOT:    [[T:%.*]] = extractelement <6 x i32> [[TMP1]], i32 0
+; CHECK-NOT:    {{%.*}} = bitcast <3 x i64> [[X:%.*]] to <6 x i32>
+; CHECK-NOT:    [[T:%.*]] = extractelement <6 x i32> {{%.*}}, i32 0
```
Comment (bader) [inline llvm/test/Transforms/InstCombine/trunc-extractelement-spir.ll]: Probably `X` and `T` are not needed as well.

Could you clarify the reasons for this change?

File: llvm/test/Transforms/OpenMP/hide_mem_transfer_latency.ll [hunk 2]
```diff
@@ -1,5 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: -p --function-signature --scrub-attributes
-; RUN: opt -S -openmp-opt-cgscc -aa-pipeline=basic-aa -openmp-hide-memory-transfer-latency < %s | FileCheck %s
+; Pass is not designed for the old pass manager -- CallGraph is not updated.
```
Comment (bader) [inline llvm/test/Transforms/OpenMP/hide_mem_transfer_latency.ll]: IIRC, @AlexeySachkov, created an issue about enabling new PM. Can you add a link to the issue here and in llvm/test/Transforms/OpenMP/values_in_offload_arrays.ll test, please?

File: llvm/test/Transforms/FunctionAttrs/2008-09-03-ReadOnly.ll [hunk 3]
```diff
@@ -1,5 +1,7 @@
-; RUN: opt < %s -basic-aa -function-attrs -S | FileCheck %s
-; RUN: opt < %s -aa-pipeline=basic-aa -passes=function-attrs -S | FileCheck %s
+; First test run uses module pass that is not designed for the old pass manager.
```
Comment (bader) [inline llvm/test/Transforms/FunctionAttrs/2008-09-03-ReadOnly.ll]: ditto

File: llvm/test/Transforms/InstCombine/trunc-extractelement-spir.ll (hunk 1 above)
Comment (vmaksimo) [reply llvm/test/Transforms/InstCombine/trunc-extractelement-spir.ll]: There was a fix to FileCheck in LLORG (fd941036bf249) - now it's not allowed to use variables inside `CHECK-NOT` checks.

File: llvm/test/Transforms/OpenMP/hide_mem_transfer_latency.ll (hunk 2 above)
Comment (vmaksimo) [reply llvm/test/Transforms/OpenMP/hide_mem_transfer_latency.ll]: @bader I can't find this issue, unfortunately. 
I hope these comments will help to fix test failures once we enable the new PM.

Comment (bader) [reply llvm/test/Transforms/OpenMP/hide_mem_transfer_latency.ll]: @AlexeySachkov, how do we track changes we did to llorg sources to disable NewPM?

Comment (AlexeySachkov) [reply llvm/test/Transforms/OpenMP/hide_mem_transfer_latency.ll]: > IIRC, @AlexeySachkov, created an issue about enabling new PM.

I didn't create such issue

**Comment by vladimirlaz** [comment]
@vmaksimo, the fix for llvm-test-suite on CUDA is available under https://github.com/intel/llvm-test-suite/pull/264
It will be submitted together with the current PR

**Pull request merged**